opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -69,70 +76,69 @@ where
 {
     fn new(config: A::InitConfig) -> Self {
         ValueMap {
-            trackers: RwLock::new(HashMap::new()),
-            has_no_attribute_value: AtomicBool::new(false),
-            no_attribute_tracker: A::create(&config),
-            count: AtomicUsize::new(0),
+            no_attribs: NoAttribs {
+                tracker: A::create(&config),
+                is_set: AtomicBool::new(false),
+            },
+            all_attribs: RwLock::new(Default::default()),
+            sorted_attribs: Mutex::new(Default::default()),
             config,
+            for_collect_after_reset: Mutex::new(Default::default()),
         }
     }
 
     fn measure(&self, value: A::PreComputedValue, attributes: &[KeyValue]) {
         if attributes.is_empty() {
-            self.no_attribute_tracker.update(value);
-            self.has_no_attribute_value.store(true, Ordering::Release);
+            self.no_attribs.tracker.update(value);
+            self.no_attribs.is_set.store(true, Ordering::Release);
             return;
         }
 
-        let Ok(trackers) = self.trackers.read() else {
-            return;
-        };
-
         // Try to retrieve and update the tracker with the attributes in the provided order first
-        if let Some(tracker) = trackers.get(attributes) {
-            tracker.update(value);
-            return;
-        }
+        match self.all_attribs.read() {
+            Ok(trackers) => {
+                if let Some(tracker) = trackers.get(attributes) {
+                    tracker.update(value);
+                    return;
+                }
+            }
+            Err(_) => return,
+        };
 
-        // Try to retrieve and update the tracker with the attributes sorted.
+        // Get or create a tracker
         let sorted_attrs = AttributeSet::from(attributes).into_vec();
-        if let Some(tracker) = trackers.get(sorted_attrs.as_slice()) {
-            tracker.update(value);
+        let Ok(mut sorted_trackers) = self.sorted_attribs.lock() else {
```

Comment:
This introduces the possibility of a update() thread fighting for the lock with *another* update() thread (or even collect()). **This is something we need to avoid as much as possible.** 
Previously, we have only read() locks, and the possibility of the update() thread getting blocked is very very small. Now that probability is quite high. 

Metrics update() should not cause a thread to be blocked - as the thread could lose its CPU slice, and need to wait for a chance to get CPU back. We never wrote this down as a key requirement, but this is an important expectation from a metrics solution.

Reply:
You probably missed, this only happens on new attribute-set, updating existing attribute set didn't changed.
In fact this PR holds locks for much less time compared to what we have in main.
It does, however, two locks in new attribute set case (sorted.lock() and all_attribs.write()), but it also has certain benefits as well. E.g. on main, when we insert new attribute-set we block readers until we finish these steps:
* get(as-is) -> (not found)
* get(sorted) -> (not found)
* create tracker
* update tracker
* insert (as-is)
* insert (sorted)
on this branch for same scenario readers are blocked until we:
* insert (as-is)

Reply:
There's several reasons for having separate hashmap:
* reduced contention on reads (as pointed out, because writes can happen mostly on separate lock)
* solves #2093 elegantly/efficiently
* allow implementing sharding efficiently. (although there is alternative impl #2305 using concurrent map, but it has aggresive locks for writes, and I have a feeling that it might be very slow for individual `.add(value, attribs)` call on metrics. Actually this `flurry` concurrent map impl, sparked me an idea to implement #2323).

I'll come with test results soon, to verify if my concerns are valid or not... :) but basically, I still hold an opinion that this is best approach to move forward.